allennlp/data/iterators/lazy_basic_iterator.py

```diff
@@ -6,13 +6,13 @@
 
 from allennlp.common import Params
 from allennlp.data.iterators.data_iterator import DataIterator
-from allennlp.data.dataset import LazyDataset, Dataset
+from allennlp.data.dataset import Batch
 from allennlp.data.instance import Instance
 
 logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
 @DataIterator.register("lazy-basic")
-class LazyBasicIterator(DataIterator[LazyDataset]):
+class LazyBasicIterator(DataIterator):
```

Comment:
Not for this PR, but will this change let us get rid of the distinction between lazy and non-lazy iterators?

Reply:
yes and no. you can see I got rid of the distinction at the type level, but e.g. `BucketIterator` calls `ensure_list` so it can shuffle its instances, and you wouldn't want to do that with a lazy dataset

Reply:
I think you can just add a parameter to `BucketIterator` that tells how many instances to pull into memory before bucketing, defaulting to the whole dataset.  You can also do that with the `BasicIterator`, before shuffling.  I _think_ that would just fix all of the issues, and we can get rid of the `Lazy` version entirely.  But, again, that's a not-too-high-priority issue for later.